pkg/sidecar/fanout/fanout_handler.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package fanout provides an http.Handler that takes in one request and fans it out to N other
+// requests, based on a list of Subscriptions. Logically, it represents all the Subscriptions to a
+// single Knative Channel.
+// It will normally be used in conjunction with multichannelfanout.Handler, which contains multiple
+// fanout.Handlers, each corresponding to a single Knative Channel.
+package fanout
+
+import (
+	"errors"
+	"github.com/knative/eventing/pkg/buses"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	"go.uber.org/zap"
+	"net/http"
+	"time"
+)
+
+const (
+	defaultTimeout = 1 * time.Minute
+
+	messageBufferSize = 500
+)
+
+// Configuration for a fanout.Handler.
+type Config struct {
+	Subscriptions []duckv1alpha1.ChannelSubscriberSpec `json:"subscriptions"`
+}
+
+// http.Handler that takes a single request in and fans it out to N other servers.
+type Handler struct {
+	config Config
+
+	receivedMessages chan *forwardMessage
+	receiver         *buses.MessageReceiver
+	dispatcher       *buses.MessageDispatcher
+	stopCh           chan<- struct{}
+
+	timeout time.Duration
+
+	logger *zap.Logger
+}
+
+var _ http.Handler = &Handler{}
+
+// forwardMessage is passed between the Receiver and the Dispatcher.
+type forwardMessage struct {
+	msg  *buses.Message
+	done chan<- error
+}
+
+// NewHandler creates a new fanout.Handler and starts a background goroutine to make it work. The
+// caller is responsible for calling handler.Stop(), when the handler is to be stopped.
+func NewHandler(logger *zap.Logger, config Config) *Handler {
+	stopCh := make(chan struct{}, 1)
+	handler := &Handler{
+		logger:           logger,
+		config:           config,
+		dispatcher:       buses.NewMessageDispatcher(logger.Sugar()),
+		receivedMessages: make(chan *forwardMessage, messageBufferSize),
+		stopCh:           stopCh,
+		timeout:          defaultTimeout,
+	}
+	// The receiver function needs to point back at the handler itself, so set it up after
+	// initialization.
+	handler.receiver = buses.NewMessageReceiver(createReceiverFunction(handler), logger.Sugar())
+
+	go handler.foreverDispatch(stopCh)
+	return handler
+}
+
+func createReceiverFunction(f *Handler) func(buses.ChannelReference, *buses.Message) error {
+	return func(_ buses.ChannelReference, m *buses.Message) error {
```

Comment:
> Call dispatch directly from the receiver function instead of using a channel. Probably the simplest, but it eliminates all backpressure since the HTTPServer will create as many goroutines as requests coming in. I recommend this strategy as the simplest to implement, most idiomatic, and least likely to surprise users. Backpressure can be layered in by a separate mechanism if necessary.

Done.